t2837: document BSD awk and ratchet hazards from t2834 in reference docs

## Task

Document three hazards surfaced during t2834 / PR #20883 in the reference docs so future shell-helper authors don't re-hit them.

## Why

These three hazards each cost a debug round-trip during PR #20883:

1. BSD awk (macOS default) silently rejects gawk-only constructs that pass linting and look correct
2. The string-literal ratchet trips on every new SVG/HTML/XML codegen helper
3. The positional-parameter ratchet baselines at zero for new files — existing helpers using `case "$1"` are grandfathered, new ones aren't

Memory recall catches these only if a worker queries the right keywords. Reference-doc rules survive context resets.

## How

Three exact-text edits across two reference docs. Worker may apply via `Edit` tool.

### Files to modify

- **EDIT:** `.agents/reference/bash-compat.md` — extend the GNU/BSD command-portability table (around line 164, the existing `awk dynamic regex` row) with two new rows for the awk hazards
- **EDIT:** `.agents/reference/shell-style-guide.md` — add a new section after the existing "Counter Safety" section (around line 158, before "Migration checklist") documenting the string-literal ratchet pattern for code-generators and the new-file ratchet behavior

### Edit 1 — bash-compat.md

**oldString:**

```text
| `awk` dynamic regex `match($0, var)` | BSD awk rejects non-literal regex | Use `gsub()` with literal patterns, or switch to `perl`. Caught production bug t1983. |
```

**newString:**

```text
| `awk` dynamic regex `match($0, var)` | BSD awk rejects non-literal regex | Use `gsub()` with literal patterns, or switch to `perl`. Caught production bug t1983. |
| `awk` 3-arg `match(str, regex, capture_array)` | BSD awk has no capture-array form (gawk-only) | Use 2-arg `match(str, regex)`, then read `RSTART`/`RLENGTH` and slice with `substr()`. Caught during t2834 / PR #20883. |
| `awk -v key="$multi_line_value"` | BSD awk rejects embedded newlines in `-v` values | Stage the multi-line content to a temp file and slurp via `BEGIN { while ((getline line < f) > 0) arr[n++] = line; close(f) }`. Caught during t2834 / PR #20883. |
```

### Edit 2 — shell-style-guide.md

**oldString:**

```text
## Migration checklist
```

**newString:**

```text
## Code-generators and the string-literal ratchet (t2834)

The string-literal ratchet (`pre-commit-hook.sh::validate_string_literals`) flags any `"..."`-quoted substring of 4+ chars that appears 3+ times in a single file, with the ratchet baselined at zero for **new** files (existing files are grandfathered at HEAD). Helpers that emit SVG, HTML, XML, or any other attribute-rich markup trip this trivially because every element repeats `width="`, `height="`, `fill="`, etc. as boundary fragments.

### Banned pattern — inline attribute fragments

Multiple `printf` calls (or one heredoc with many elements) each containing inline `attr="value"` fragments:

```bash
# Trips the ratchet — `" fill="`, `" width="`, `" height="` count >= 3
printf '  <rect x="%s" y="%s" width="%s" height="%s" fill="%s"/>\n' "$x" "$y" "$w" "$h" "$c"
printf '  <text x="%s" y="%s" font-family="%s" fill="%s">%s</text>\n' "$x" "$y" "$f" "$c" "$t"
```

### Allowed pattern — attribute-builder helpers

Build attribute strings via a single `'%s="%s" '` format template, eliminating inline fragments entirely:

```bash
_ATTR_FMT='%s="%s" '

_svg_attrs() {
    local _out=""
    while [[ $# -ge 2 ]]; do
        local _k="$1" _v="$2"
        # shellcheck disable=SC2059  # _ATTR_FMT is a trusted constant
        _out+=$(printf "$_ATTR_FMT" "$_k" "$_v")
        shift 2
    done
    printf '%s' "${_out%% }"
    return 0
}

_svg_elem() {
    local _tag="$1"; shift
    local _attrs; _attrs=$(_svg_attrs "$@")
    printf '  <%s %s/>\n' "$_tag" "$_attrs"
    return 0
}

# Usage — no inline attr= fragments, no ratchet trip
_svg_elem rect x "$x" y "$y" width "$w" height "$h" fill "$c"
```

Canonical implementation: `.agents/scripts/loc-badge-helper.sh::_svg_attrs / _svg_elem / _svg_open / _svg_close / _svg_text_elem`.

### Ratchet baseline-at-zero for new files

Every ratchet-style validator (string-literal, positional-parameter, function-complexity, nesting-depth, file-size) compares HEAD content vs staged content. For an **existing** file with pre-existing violations, the ratchet only blocks when the staged count exceeds HEAD — pre-existing debt is grandfathered.

For a **new** file, HEAD count is zero. Every violation is "new" and blocks the commit.

**Practical rule:** in a new file, follow the strict pattern from line 1. The "fix later" approach available to maintenance commits on legacy files is not available to new files.

Examples of strict patterns required from line 1 of every new shell helper:

- Positional parameters: `local _arg="$1"` always (never bare `case "$1" in` or `VAR="$2"`)
- Function arguments: `local _msg="$1"` at the top of every function body
- Repeated string fragments: extract to constants OR build via a helper template (see attribute-builder pattern above)

### Detection

Reproduce the validator locally before committing a new helper:

```bash
# Show literals that would trigger the ratchet (after the canonical sed pre-strip)
grep -v '^[[:space:]]*#' your-helper.sh | sed -E '
  s/"\$[A-Za-z_][A-Za-z0-9_]*"//g
  s/"\$\{[^}]*\}"//g
  s/"\$@"//g
  s/"\$[0-9*#?$!-]"//g
' | grep -oE '"[^"]{4,}"' | grep -vE '^"[0-9]+\.?[0-9]*"$' | grep -vE '^"\$' | sort | uniq -c | awk '$1 >= 3' | sort -rn

# Show direct positional-parameter usage that would trigger
awk '
  { line = $0
    gsub(/\047[^\047]*\047/, "", line)
    if (line ~ /^[[:space:]]*#/) next
    sub(/[[:space:]]+#.*/, "", line)
    if (line ~ /local[[:space:]].*=.*\$[1-9]/) next
    if (line ~ /\$[1-9]/) print NR ": " $0
  }
' your-helper.sh
```

Both must be empty before commit.

## Migration checklist
```

### Verification

```bash
# Files exist after edit
test -f .agents/reference/bash-compat.md
test -f .agents/reference/shell-style-guide.md

# bash-compat has the two new rows
grep -c "3-arg \`match(str, regex, capture_array)\`" .agents/reference/bash-compat.md  # expect 1
grep -c "BSD awk rejects embedded newlines" .agents/reference/bash-compat.md  # expect 1

# shell-style-guide has the new section
grep -c "Code-generators and the string-literal ratchet (t2834)" .agents/reference/shell-style-guide.md  # expect 1
grep -c "Ratchet baseline-at-zero for new files" .agents/reference/shell-style-guide.md  # expect 1

# Markdown still lints clean
npx --yes markdownlint-cli2 .agents/reference/bash-compat.md .agents/reference/shell-style-guide.md
```

## Acceptance

- [ ] `bash-compat.md` table extended with two new awk-portability rows
- [ ] `shell-style-guide.md` has a new "Code-generators and the string-literal ratchet (t2834)" section before the existing "Migration checklist" section
- [ ] Both files lint clean with `markdownlint-cli2`
- [ ] No other files modified (Files Scope is exactly these two)

## Session Origin

interactive — followup of t2834 / PR #20883

## Files Scope

- `.agents/reference/bash-compat.md`
- `.agents/reference/shell-style-guide.md`

## Context

- Tier rationale: `tier:simple` because the brief contains verbatim oldString/newString blocks, target files are under 500 lines, and there's no design judgment — just append two table rows and one section.
- The exact patterns to document are crystallized in this issue body; the worker should not re-invent or paraphrase. Apply the diffs as written.

<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.11.1 plugin for [OpenCode](https://opencode.ai) v1.14.25 with claude-opus-4-7 spent 46m and 127,203 tokens on this with the user in an interactive session.

--- a/.agents/reference/bash-compat.md
+++ b/.agents/reference/bash-compat.md
@@ -162,6 +162,8 @@ Even with a compatible bash version, external commands diverge silently between 
 | `sed -r '...'` (extended regex) | `sed -E '...'` | `sed -E` works on both — always use `-E`, never `-r`. |
 | `grep -P '...'` (PCRE) | N/A on BSD grep | Use `grep -E` with POSIX ERE, or pipe to `perl -ne '...'`. |
 | `awk` dynamic regex `match($0, var)` | BSD awk rejects non-literal regex | Use `gsub()` with literal patterns, or switch to `perl`. Caught production bug t1983. |
+| `awk` 3-arg `match(str, regex, capture_array)` | BSD awk has no capture-array form (gawk-only) | Use 2-arg `match(str, regex)`, then read `RSTART`/`RLENGTH` and slice with `substr()`. Caught during t2834 / PR #20883. |
+| `awk -v key="$multi_line_value"` | BSD awk rejects embedded newlines in `-v` values | Stage the multi-line content to a temp file and slurp via `BEGIN { while ((getline line < f) > 0) arr[n++] = line; close(f) }`. Caught during t2834 / PR #20883. |
 | `xargs -r` | BSD xargs has no `-r`; empty input is a no-op by default | Pre-check input with `[[ -s input ]]` before calling xargs, or use `find ... -exec {} +` which is portable. |
 | `find -printf` | BSD find has no `-printf` | Use `find ... -exec <fmt-cmd> {} +` or pipe through `stat`. |
 | `mktemp --suffix=.X` | BSD mktemp uses `-t prefix` differently | `mktemp -t aidevops.XXXXXX` works on both; never use `--suffix`. |

--- a/.agents/reference/shell-style-guide.md
+++ b/.agents/reference/shell-style-guide.md
@@ -157,6 +157,93 @@ The scanner skips any file with that directive.
 - Parent issue #20581 (t2762) — systemic sweep and prevention
 - Canonical reference implementation: `.agents/scripts/progressive-load-check.sh:80-97` (pre-existing correct counter usage)
 
+## Code-generators and the string-literal ratchet (t2834)
+
+The string-literal ratchet (`pre-commit-hook.sh::validate_string_literals`) flags any `"..."`-quoted substring of 4+ chars that appears 3+ times in a single file, with the ratchet baselined at zero for **new** files (existing files are grandfathered at HEAD). Helpers that emit SVG, HTML, XML, or any other attribute-rich markup trip this trivially because every element repeats `width="`, `height="`, `fill="`, etc. as boundary fragments.
+
+### Banned pattern — inline attribute fragments
+
+Multiple `printf` calls (or one heredoc with many elements) each containing inline `attr="value"` fragments:
+
+```bash
+# Trips the ratchet — `" fill="`, `" width="`, `" height="` count >= 3
+printf '  <rect x="%s" y="%s" width="%s" height="%s" fill="%s"/>\n' "$x" "$y" "$w" "$h" "$c"
+printf '  <text x="%s" y="%s" font-family="%s" fill="%s">%s</text>\n' "$x" "$y" "$f" "$c" "$t"
+```
+
+### Allowed pattern — attribute-builder helpers
+
+Build attribute strings via a single `'%s="%s" '` format template, eliminating inline fragments entirely:
+
+```bash
+_ATTR_FMT='%s="%s" '
+
+_svg_attrs() {
+    local _out=""
+    while [[ $# -ge 2 ]]; do
+        local _k="$1" _v="$2"
+        # shellcheck disable=SC2059  # _ATTR_FMT is a trusted constant
+        _out+=$(printf "$_ATTR_FMT" "$_k" "$_v")
+        shift 2
+    done
+    printf '%s' "${_out%% }"
+    return 0
+}
+
+_svg_elem() {
+    local _tag="$1"; shift
+    local _attrs; _attrs=$(_svg_attrs "$@")
+    printf '  <%s %s/>\n' "$_tag" "$_attrs"
+    return 0
+}
+
+# Usage — no inline attr= fragments, no ratchet trip
+_svg_elem rect x "$x" y "$y" width "$w" height "$h" fill "$c"
+```
+
+Canonical implementation: `.agents/scripts/loc-badge-helper.sh::_svg_attrs / _svg_elem / _svg_open / _svg_close / _svg_text_elem`.
+
+### Ratchet baseline-at-zero for new files
+
+Every ratchet-style validator (string-literal, positional-parameter, function-complexity, nesting-depth, file-size) compares HEAD content vs staged content. For an **existing** file with pre-existing violations, the ratchet only blocks when the staged count exceeds HEAD — pre-existing debt is grandfathered.
+
+For a **new** file, HEAD count is zero. Every violation is "new" and blocks the commit.
+
+**Practical rule:** in a new file, follow the strict pattern from line 1. The "fix later" approach available to maintenance commits on legacy files is not available to new files.
+
+Examples of strict patterns required from line 1 of every new shell helper:
+
+- Positional parameters: `local _arg="$1"` always (never bare `case "$1" in` or `VAR="$2"`)
+- Function arguments: `local _msg="$1"` at the top of every function body
+- Repeated string fragments: extract to constants OR build via a helper template (see attribute-builder pattern above)
+
+### Detection
+
+Reproduce the validator locally before committing a new helper:
+
+```bash
+# Show literals that would trigger the ratchet (after the canonical sed pre-strip)
+grep -v '^[[:space:]]*#' your-helper.sh | sed -E '
+  s/"\$[A-Za-z_][A-Za-z0-9_]*"//g
+  s/"\$\{[^}]*\}"//g
+  s/"\$@"//g
+  s/"\$[0-9*#?$!-]"//g
+' | grep -oE '"[^"]{4,}"' | grep -vE '^"[0-9]+\.?[0-9]*"$' | grep -vE '^"\$' | sort | uniq -c | awk '$1 >= 3' | sort -rn
+
+# Show direct positional-parameter usage that would trigger
+awk '
+  { line = $0
+    gsub(/\047[^\047]*\047/, "", line)
+    if (line ~ /^[[:space:]]*#/) next
+    sub(/[[:space:]]+#.*/, "", line)
+    if (line ~ /local[[:space:]].*=.*\$[1-9]/) next
+    if (line ~ /\$[1-9]/) print NR ": " $0
+  }
+' your-helper.sh
+```
+
+Both must be empty before commit.
+
 ## Migration checklist
 
 1. Identify current pattern (plain, readonly, include-guard, or prefixed).

